modules/3d/test/test_pose_graph.cpp

```diff
@@ -155,6 +155,136 @@ TEST(PoseGraph, sphereG2O)
+                        .setMaxIterations(100)
+                        .setCheckRelEnergyChange(true)
+                        .setRelEnergyDeltaTolerance(1e-6)
+                        .setGeodesic(true));
+
+    auto r1 = pgWithMSTAndOptimizer->optimize();
+    auto r2 = pgOptimizerOnly->optimize();
+
+    EXPECT_TRUE(r1.found);
+    EXPECT_TRUE(r2.found);
+    EXPECT_LE(r2.energy, 1.47723e+06);
+    // Allow small tolerance due to optimization differences; final energy/iterations are effectively the same
+    EXPECT_LE(std::abs(r1.energy - r2.energy), 1e-2);
+    ASSERT_LE(std::abs(r1.iters - r2.iters), 1);
+
+    // Add the "--test_debug" to arguments to see resulting pose graph nodes positions
+    if (cvtest::debugLevel > 0)
+    {
+        // Write OBJ for MST-initialized pose graph with optimizer
+        std::string fname = "pg_with_mst_and_optimizer.obj";
+        std::fstream of(fname, std::fstream::out);
+        std::vector<size_t> ids = pgWithMSTAndOptimizer->getNodesIds();
+        for (const size_t& id : ids)
+        {
+            Point3d d = pgWithMSTAndOptimizer->getNodePose(id).translation();
+            of << "v " << d.x << " " << d.y << " " << d.z << std::endl;
+        }
+        size_t esz = pgWithMSTAndOptimizer->getNumEdges();
+        for (size_t i = 0; i < esz; i++)
+        {
+            size_t sid = pgWithMSTAndOptimizer->getEdgeStart(i), tid = pgWithMSTAndOptimizer->getEdgeEnd(i);
+            of << "l " << sid + 1 << " " << tid + 1 << std::endl;
+        }
+        of.close();
+
+        // Write OBJ for optimizer-only pose graph
+        std::string fname = "pg_optimizer_only.obj";
+        std::fstream of(fname, std::fstream::out);
+        std::vector<size_t> ids = pgOptimizerOnly->getNodesIds();
+        for (const size_t& id : ids)
+        {
+            Point3d d = pgOptimizerOnly->getNodePose(id).translation();
+            of << "v " << d.x << " " << d.y << " " << d.z << std::endl;
+        }
+        size_t esz = pgOptimizerOnly->getNumEdges();
+        for (size_t i = 0; i < esz; i++)
+        {
+            size_t sid = pgOptimizerOnly->getEdgeStart(i), tid = pgOptimizerOnly->getEdgeEnd(i);
+            of << "l " << sid + 1 << " " << tid + 1 << std::endl;
+        }
+        of.close();
+
+        // Write OBJ for initial pose graph
+        std::string fname = "pg_init.obj";
+        std::fstream of(fname, std::fstream::out);
+        std::vector<size_t> ids = init->getNodesIds();
+        for (const size_t& id : ids)
+        {
+            Point3d d = init->getNodePose(id).translation();
+            of << "v " << d.x << " " << d.y << " " << d.z << std::endl;
+        }
+        size_t esz = init->getNumEdges();
+        for (size_t i = 0; i < esz; i++)
+        {
+            size_t sid = init->getEdgeStart(i), tid = init->getEdgeEnd(i);
+            of << "l " << sid + 1 << " " << tid + 1 << std::endl;
+        }
+        of.close();
```

Comment:
Looks like this code is not needed any more, as you added implementation with `saveMesh` below.